src/MongoDB.Driver.Core/Core/Authentication/MongoAWSAuthenticator.cs

```diff
@@ -0,0 +1,539 @@
+            string username,
+            IEnumerable<KeyValuePair<string, string>> properties,
+            IRandomByteGenerator randomByteGenerator,
+            IClock clock)
+            : base(CreateMechanism(username, null, properties, randomByteGenerator, clock))
+        {
+        }
+
+        /// <inheritdoc/>
+        public override string DatabaseName
+        {
+            get { return "$external"; }
+        }
+
+        // private static methods
+        private static MongoAWSMechanism CreateMechanism(
+            UsernamePasswordCredential credential,
+            IEnumerable<KeyValuePair<string, string>> properties,
+            IRandomByteGenerator randomByteGenerator,
+            IClock clock)
+        {
+            if (credential.Source != "$external")
+            {
+                throw new ArgumentException("MONGODB-AWS authentication may only use the $external source.", nameof(credential));
+            }
+
+            return CreateMechanism(credential.Username, credential.Password, properties, randomByteGenerator, clock);
+        }
+
+        private static MongoAWSMechanism CreateMechanism(
+            string username,
+            SecureString securePassword,
+            IEnumerable<KeyValuePair<string, string>> properties,
+            IRandomByteGenerator randomByteGenerator,
+            IClock clock)
+        {
+            ValidateMechanismProperties(properties);
+
+            var awsCredentials =
+                CreateAwsCredentialsFromMongoCredentials(username, securePassword, properties) ??
+                CreateAwsCredentialsFromEnvironmentVariables() ??
+                CreateAwsCredentialsFromEcsResponse() ??
+                CreateAwsCredentialsFromEc2Response();
+
+            if (awsCredentials == null)
+            {
+                throw new InvalidOperationException("A MONGODB-AWS credential must have an access key ID.");
+            }
+
+            var credentials = new UsernamePasswordCredential("$external", awsCredentials.Username, awsCredentials.Password);
+
+            return new MongoAWSMechanism(credentials, awsCredentials.SessionToken, randomByteGenerator, clock);
+        }
+
+        private static AwsCredentials CreateAwsCredentialsFromMongoCredentials(string username, SecureString securePassword, IEnumerable<KeyValuePair<string, string>> properties)
+        {
+            var sessionToken = ExtractSessionTokenFromMechanismProperties(properties);
+
+            var awsCredentials = new AwsCredentials
+            {
+                Username = username,
+                Password = securePassword,
+                SessionToken = sessionToken
+            };
+
+            ValidateAwsCredentials(awsCredentials);
+
+            if (awsCredentials.Username == null)
+            {
+                return null;
+            }
+            
+            return awsCredentials;
+        }
+
+        private static AwsCredentials CreateAwsCredentialsFromEnvironmentVariables()
```

Comment:
Refactor this method to:

    private static AwsCredentials CreateAwsCredentialsFromEnvironmentVariables()
    {
        var accessKeyId = Environment.GetEnvironmentVariable("AWS_ACCESS_KEY_ID");
        var secretAccessKey = Environment.GetEnvironmentVariable("AWS_SECRET_ACCESS_KEY");
        var sessionToken = Environment.GetEnvironmentVariable("AWS_SESSION_TOKEN");

        if (accessKeyId == null && secretAccessKey == null && sessionToken == null)
        {
            return null;
        }
        if (secretAccessKey != null && accessKeyId == null)
        {
            throw new InvalidOperationException("When using MONGODB-AWS authentication if a secret access key is provided via environment variables then an access key ID must be provided also.");
        }
        if (accessKeyId != null && secretAccessKey == null)
        {
            throw new InvalidOperationException("When using MONGODB-AWS authentication if an access key ID is provided via environment variables then a secret access key must be provided also.");
        }
        if (sessionToken != null && (accessKeyId == null || secretAccessKey == null))
        {
            throw new InvalidOperationException("When using MONGODB-AWS authentication if a session token is provided via environment variables then an access key ID and a secret access key must be provided also.");
        }

        return new AwsCredentials(accessKeyId, ToSecureString(secretAccessKey), sessionToken);
    }